src/realm/object-store/sync/sync_session.cpp

```diff
@@ -977,40 +977,24 @@ void SyncProgressNotifier::set_local_version(uint64_t snapshot_version)
 std::function<void()> SyncProgressNotifier::NotifierPackage::create_invocation(Progress const& current_progress,
                                                                                bool& is_expired)
 {
-    uint64_t transferrable;
-    if (is_streaming) {
-        transferrable = is_download ? current_progress.downloadable : current_progress.uploadable;
-    }
-    else if (captured_transferrable) {
-        if (is_download && (current_progress.downloadable == 0)) {
-            captured_transferrable = 0;
-        }
-        transferrable = *captured_transferrable;
-    }
-    else {
-        if (is_download) {
-            // In protocol versions 25 and earlier, downloadable_bytes was the total
-            // size of the history. From protocol version 26, downloadable_bytes
-            // represent the non-downloaded bytes on the server. Since the user supplied
-            // progress handler interprets downloadable_bytes as the total size of
-            // downloadable bytes, this number must be calculated.  We could change the
-            // meaning of downloadable_bytes for the progress handler, but that would be
-            // a breaking change. Note that protocol version 25 (and earlier) is no
-            // longer supported by clients.
-            captured_transferrable = current_progress.downloaded + current_progress.downloadable;
-        }
-        else {
-            // If the sync client has not yet processed all of the local
-            // transactions then the uploadable data is incorrect and we should
-            // not invoke the callback
-            if (snapshot_version > current_progress.snapshot_version)
-                return [] {};
-            captured_transferrable = current_progress.uploadable;
-        }
+    uint64_t transferred = is_download ? current_progress.downloaded : current_progress.uploaded;
+    uint64_t transferrable = is_download ? current_progress.downloadable : current_progress.uploadable;
+    if (!is_streaming) {
+        // If the sync client has not yet processed all of the local
+        // transactions then the uploadable data is incorrect and we should
+        // not invoke the callback
+        if (!is_download && snapshot_version > current_progress.snapshot_version)
+            return [] {};
+
+        // The initial download size we get from the server is an estimate
+        // and it may decrease once compaction is performed, so we need to
```

Comment:
This is not true. Once compaction is performed, the downloadable bytes can be greater than the actual number of bytes available, and so we'll never reach it. The only guarantee the server makes about downloadable_bytes is that it will be zero when there are no more download messages to send. That's what the original PR was trying to address.

Reply:
I don't understand how that's different? Receiving downloadable_bytes=0 from the server makes the value we get here go down.

Reply:
I suppose we could special-case getting downloadable_bytes=0 rather than just handling any values which has shrunk by more than downloaded_bytes increased by, but that just seems worse and then we'd need error-handling for the case where it decreased by too much but not to zero.

Reply:
In practice, you're right, but this is not the guarantee the server actually gives us. The only guarantee we get is that downloadable_bytes will be zero when there are no more download messages to send.

Reply:
If the contract with the server is that it sends a random value in downloadable_bytes if there is more to download and zero if there is not, then the server has made it impossible to implement non-streaming progress notifiers (or useful progress notifiers at all), and the protocol is completely insane.